torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+
+    def __init__(
+        self,
+        wrapped_function: WrappedFunction,
+        id: GraphID,
+        parent: Optional[CUDAGraphNode],
+        inputs: List[Tensor],
+        cuda_graphs_pool: Tuple[int, int],
+    ):
+        assert isinstance(inputs, (list, tuple))
+
+        self.wrapped_function = wrapped_function
+        self.id = id
+
+        # if this is a root parent will be None. use weakref to prevent reference cycle
+        self._parent = weakref.ref(parent) if parent is not None else None
+        self.cuda_graphs_pool = cuda_graphs_pool
+
+        # A single wrapped function may be recorded multiple times if memory patterns or
+        # invariants change from one execution to the next
+        self.children: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
+
+        self.device: int = next(
+            (x.device.index for x in inputs if is_cuda_tensor), None
+        )
+
+        # we preserve a single reference to executed outputs that is then referenced
+        # in children to avoid children having to chase parent pointers in the hot path
+        # DO NOT reassign output_weakrefs, only call `clear()`
+        # Path is a series of nodes from root to the current node
+        self.outputs_weakrefs: List[Optional[TensorWeakRef]] = []
+        self.path_weakrefs: List[List[Optional[TensorWeakRef]]] = [
+            node.outputs_weakrefs for node in self._path_from_root
+        ]
+
+        # tensors which are outputs of previous graphs in the tree
+        self.cudagraph_managed_idxs: List[int] = [
+            idx
+            for idx, t in enumerate(inputs)
+            if self._is_cuda_graph_recorded_tensor(t)
+        ]
+
+        self.static_input_idxs: List[int] = list(
+            set(wrapped_function.static_input_idxs) | set(self.cudagraph_managed_idxs)
+        )
+
+        self.static_input_data_ptrs: List[int] = [
+            (inputs[i].data_ptr() if i in self.static_input_idxs else None)
+            for i in range(len(inputs))
+        ]
+        self.expanded_dims: List[List[int]] = [
+            get_expanded_dims(x) if idx not in self.static_input_idxs else []
+            for idx, x in enumerate(inputs)
+        ]
+
+        # For each node in path, which outputs were observed to be live
+        # before invoking graph recording, and after graph recording
+        self.recorded_liveness_before_graph: List[List[bool]] = []
+        self.recorded_liveness_after_graph: List[List[bool]] = []
+
+        # indices into node at level i for output j
+        self.expected_dead_indices_before_graph: List[Tuple[int, int]] = []
+        self.expected_dead_indices_after_graph: List[Tuple[int, int]] = []
+
+        # all live indices after graph recording
+        self.live_indices_after_graph: List[Tuple[int, int]] = []
+
+        if self.parent is not None:
+            previous_liveness = self.parent.recorded_liveness_after_graph
+            curr_liveness = self._get_liveness(self.path_weakrefs)
```

Comment:
If these methods don't actually depend on entries in self, I might suggest declaring them as top level functions to make it clear they don't access self in an unrestricted way.